c/misra/test/rules/RULE-18-8/test.c

```diff
@@ -1,15 +1,37 @@
 #define TEST 1
 
-void f(int n, int pa[1][n]) { // NON_COMPLIANT
-  int a[1];                   // COMPLIANT
+void f(int n) {
+  int a[1]; // COMPLIANT
   int x = 1;
   int a1[1 + x];  // NON_COMPLIANT - not integer constant expr
   int a2[n];      // NON_COMPLIANT
   int a3[1][n];   // NON_COMPLIANT
   int a4[] = {1}; // COMPLIANT - not a VLA
   int a5[TEST];   // COMPLIANT
   int a6[1 + 1];  // COMPLIANT
+  int a7[n][1];   // NON_COMPLIANT
+  int(*a8)[n];    // COMPLIANT - pointer to VLA, see RULE-18-10
+
+  extern int e1[]; // COMPLIANT
+
+  // A typedef is not a VLA. However, `VlaDeclStmt`s match the typedef.
+  typedef int vlaTypedef[n]; // COMPLIANT[FALSE_POSITIVE]
+  vlaTypedef t1;             // NON_COMPLIANT[FALSE_NEGATIVE]
```

Comment:
I'm happy to do it that way, and I think the first point is a good reason why we would want to do it this way. That said, in my conversations with Robert, our conclusion was that 18-8 is intended to reject allocating VLAs. It's the allocation of a VLA that makes them so dangerous to use. In my head, a VLA is an object in memory, while a VLA type is the static type of that object, and the rule now states that it's VLAs which are banned, not VLA types.

Of course, an unused VLA typedef should be deleted. And most uses of a VLA typedef would be invalid. But in theory, not all. A VLA typedef could be used in a `sizeof()`, or it could be used in a context where it is adjusted to a pointer.

I don't think these cases are likely to come up, I did want to document them correctly in the tests.

Perhaps I should add this explanation in a comment?

Reply:
Re-reading the rule, I think your interpretation as it intending to identify allocations is correct. The rule starts by saying:
> Variable-length arrays are specified when the size of an array declared in a block or function prototype is not an integer constant expression.

Which would exclude an unused typedef, or a typedef that is not used to declare an array in a block or function prototype. This was a change from the previous interpretation for Rule 18.8 in MISRA C 2012.

I think there's two options for how we handle this:
 1. We continue to flag the typedef, but only report it if we find it used as the type in a variable declaration in a "valid" location - e.g. where there exists a `Variable` whose type is the typedef (or derived from the typedef?). We could add the `Variable`s as placeholders in this case.
 2. We re-jig the query to report `Variable`s, where the `Variable` is the `getVariable()` of the `VlaDeclStmt` or where the variable type is a use of a the `TypeDef` associated with a `VlaDeclStmt`.

If you've used a `typedef` I think it's likely you're intending to use the type in multiple places, so I would say that's an argument to continue to flag the `typedef`. However, it may not be that common in practice, so it's possibly a premature optimization in reporting.

Reply:
Resolved via 2!